bls/algebra/src/bls/keys.rs

```diff
@@ -177,6 +223,22 @@ impl PublicKey {
     }
 }
 
+impl PartialEq for PublicKey {
+    fn eq(&self, other: &Self) -> bool {
+        // The equality operator in G2Projective is very slow
+        self.pk.x == other.pk.x && self.pk.y == other.pk.y && self.pk.z == other.pk.z
```

Comment:
Are we guaranteed the public key will be in a canonical form? This will be wrong if it takes in two different projective coordinates representing the same equivalence class.

Reply:
1) All the keys using equal are coming directly from deserialize, so they should be in the same form
2) It is being used for caching only, so a false negative on equal just means we recompute (and in the case the slower equal operator is so expensive recomputing is better)
3) Another equal would violate the requirements of HashSet described above.  (`k1 == k2` must imply `hash(k1) == hash(k2)`)

Reply:
Maybe add a comment describing this?